Brace initialization for simple POD classes

wjakob opened the pull request:
For simple POD classes storing just a few fields, I often don't bother writing constructors and use the generalized C++11 brace initialization syntax. pybind11 doesn't like that and currently requires one to add constructors to the POD class (the ``py::init`` details use the C++03 syntax). This commit switches the pybind11 implementation to use brace initialization as well.

^@jagerman
^@dean0x7d
Unrelated: I almost did a force push to the master branch instead of my personal repo but noticed in the last moment. I switched the master branch to protected status now (i.e. force pushes not allowed) to prevent accidents -- it's easily removed temporarily if we ever need to.

wjakob commented:
done.

jagerman commented:
> Unrelated: I almost did a force push to the master branch instead of my personal repo but noticed in the last moment.

I deliberately keep the `pybind11/master` remote named locally as `upstream` (really anything other than `master` will work fine).  The main advantage is that `git` won't push to it by default because of the branch naming mismatch, instead giving the error:
```C++
fatal: The upstream branch of your current branch does not match
the name of your current branch.  To push to the upstream branch
on the remote, use

    git push upstream HEAD:master

To push to the branch of the same name on the remote, use

    git push upstream upstream
```
so that I can still push it using the first suggested command, but it's always an extra safety step (even against ordinary pushes, not just forced pushes).

jagerman commented on docs/advanced/classes.rst:
```diff
@@ -490,6 +492,26 @@ you could equivalently write:
+        .def(py::init<int, const std::string &>());
+
+Note that brace initialization preferentially invokes constructor overloads
+taking a ``std::initializer_list``.
```
Perhaps add:

In the rare event that this causes an issue, you can work around it by using `py::init(...)` with a lambda function that constructs the new object as desired.

Since that makes it more than just a single sentence, it's probably useful to put it in a `note` section.

jagerman commented:
One small suggested addition to the docs, but otherwise this LGTM.

dean0x7d commented on docs/advanced/classes.rst:
```diff
@@ -490,6 +492,26 @@ you could equivalently write:
+
+.. code-block:: cpp
+
+    struct POD {
```
Not a POD (`std::string` is not trivially copyable). `Aggregate` should be the term for this kind of structure and brace initialization.

wjakob commented:
Rebased and squashed down into 2 commits.

Unrelated: I noticed a number of warnings when compiling the test suite on Clang/OSX -- at least one of them is related to the new factory constructors.

```
[9/9/17] Building CXX object tests/CMakeFiles/pybind11_tests.dir/test_factory_constructors.cpp.o
/Users/wjakob/projects/pybind11/tests/test_factory_constructors.cpp:180:17: warning: lambda capture 'c' is not used [-Wunused-lambda-capture]
    auto c4a = [c](pointer_tag, TF4_tag, int a) { return new TestFactory4(a);};
                ^
1 warning generated.
[7/9/19] Building CXX object tests/CMakeFiles/pybind11_tests.dir/test_numpy_array.cpp.o
In file included from /Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:12:
include/pybind11/numpy.h:394:40: warning: implicit conversion changes signedness: 'long' to 'unsigned long' [-Wsign-conversion]
        static_assert(sizeof...(Ix) == Dims || Dynamic,
                                    ~~ ^~~~
/Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:239:18: note: in instantiation of function template specialization 'pybind11::detail::unchecked_mutable_reference<double, -1>::operator()<long, long>' requested here
                r(i, j) += v;
                 ^
In file included from /Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:12:
include/pybind11/numpy.h:345:40: warning: implicit conversion changes signedness: 'long' to 'unsigned long' [-Wsign-conversion]
        static_assert(sizeof...(Ix) == Dims || Dynamic,
                                    ~~ ^~~~
include/pybind11/numpy.h:396:43: note: in instantiation of function template specialization 'pybind11::detail::unchecked_reference<double, -1>::operator()<long, long>' requested here
        return const_cast<T &>(ConstBase::operator()(index...));
                                          ^
/Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:239:18: note: in instantiation of function template specialization 'pybind11::detail::unchecked_mutable_reference<double, -1>::operator()<long, long>' requested here
                r(i, j) += v;
                 ^
In file included from /Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:12:
include/pybind11/numpy.h:394:40: warning: implicit conversion changes signedness: 'long' to 'unsigned long' [-Wsign-conversion]
        static_assert(sizeof...(Ix) == Dims || Dynamic,
                                    ~~ ^~~~
/Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:248:14: note: in instantiation of function template specialization 'pybind11::detail::unchecked_mutable_reference<double, -1>::operator()<long, long, long>' requested here
            r(i, j, k) = start++;
             ^
In file included from /Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:12:
include/pybind11/numpy.h:345:40: warning: implicit conversion changes signedness: 'long' to 'unsigned long' [-Wsign-conversion]
        static_assert(sizeof...(Ix) == Dims || Dynamic,
                                    ~~ ^~~~
include/pybind11/numpy.h:396:43: note: in instantiation of function template specialization 'pybind11::detail::unchecked_reference<double, -1>::operator()<long, long, long>' requested here
        return const_cast<T &>(ConstBase::operator()(index...));
                                          ^
/Users/wjakob/projects/pybind11/tests/test_numpy_array.cpp:248:14: note: in instantiation of function template specialization 'pybind11::detail::unchecked_mutable_reference<double, -1>::operator()<long, long, long>' requested here
            r(i, j, k) = start++;
```

dean0x7d commented:
> Unrelated: I noticed a number of warnings when compiling the test suite on Clang/OSX

Which version of clang? I'm not seeing these with appleclang v8.1.0. (In fact, seems to be reporting `unknown warning option '-Wunused-lambda-capture'`).

wjakob commented:
It's clang trunk compiled from SVN :).

wjakob commented:
merged.
